### PR TITLE
[i18n] Add BCP47 Language Tags Parsing

### PR DESCRIPTION
**Pull request by CrystalP**
## Description

Implement a BCP 47 language tags parser with full support of RFC5646, a fast path for tags made up of the language subtag only (ex: en, ara), and convert a few code paths to use the new functionality.

### Language descriptions

The English language description displayed for subtitles and audio tracks now supports bcp47 language tags and mimicks locale names. It is however limited by the lack of support of the IANA language subtags registry and for now only the recently updated existing ISO639-1/-2 and ISO 3166-1 support is used to convert languages and regions to their English names.

For example:
- en-AU => English (Australia) (en is ISO 639-1 and AU is ISO 3166-1)
- es-419 => Spanish (419) (es is ISO 639-1, 419 is M.49 and requires the registry)
- zh-yue-Hant-HK => Chinese (yue Hant Hong Kong) (extlang yue and script Hant require the registry)
- yue-HK => yue (Hong Kong) (yue is ISO 639-3 and requires the registry)

As was already happening before the PR, user languages defined in as.xml or language addons take precedence and provide "friendlier" names.

### External subtitles / audio tracks

Modified the parsing of external subtitles /audio tracks to accept bcp47 language tags in the filename. The - separators have to be replaced with _ since - is already a segment separator for filenames. For example:
Big.Buck.Bunny.forced.**en_AU**.srt

### Movies and TV Shows original language

The original language scraped or imported from nfo is now stored as BCP47 in the video db. Existing entries were converted from ISO 639-2/B to BCP47 by a DB bump.

The original language field was missed in a couple places of the VideoInfoTag when it was added and this is now corrected.

### Other support of BCP47 in media

There is nothing to change for InputStream as far as I can tell, it already forwards bcp47 language tags to core and benefits from the improved conversion to English language names of the PR.

Internal tracks of file-based media are problematic due to missing support of BCP47 in ffmpeg, at least for mkv (mp4 seems problematic as well).
The workaround of entering the language tag in the track name, surrounded by curly braces, was upgraded to the new BCP 47 parser.

### Future PR
- a fast path could be added for language+region or language+script if it is felt necessary, in order to avoid the full blown regexp parsing for common/easy tags
- implement the match algorithms of BCP47 to improve the selection of the best audio or subtitle track to
- implement the IANA language subtags registry for better conversion of BCP47 to English and for validation of the tags
- internally store track languages as BCP47 objects instead of string to reduce the amount of parsing
- alternatively implement caching of parsed tags
- what to do about the translation of language names themselves - the registry or ISO norms only contain English names for all subtags

## Motivation and context

Many code paths were limited to 3 letter ISO 639-2 codes, not sufficient to express the variety of languages.
BCP 47 is the standard.

## How has this been tested?

Added unit tests for everything. All pass.
Confirmed that external subs, mkv, ia have the additional subtags, and that those are translated to English on screen (within limits imposed by lack of support of the IANA language subtags registry.

## What is the effect on users?

more languages recognized without having to define entries in advancedsettings.xml/language codes or requiring language addons to be installed.

## Screenshots (if appropriate):
external subs
image

inputstream adaptative
image

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [X] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [X] I have added tests to cover my change
- [X] All new and existing tests passed

### CODE REVIEW

**Comment by garbear** [comment]
Great job getting up the final PR! @claude guide human eyes on how to best review this PR.

**Comment by CrystalP** [comment]
Give me a minute to ponder all that.
I can already see that the critical issues are not valid because already addressed or not in scope of the PR for 2. Logic Error (xbmc/utils/LangCodeExpander.cpp:239)

**Comment by CrystalP** [comment]
Replying to the comments of Claude:
- Database migration: needed some finetuning - done.
- Parser regexp correctness/edge cases: verified by many unit tests - added more
- The validity of BCP 47 tags cannot be implemented without the registry, which cannot happen in this PR as it would make it a monster. I implemented what could be given the ISO data already available.
-  Integration Points in LangCodeExpander: unit tests for the various inputs - added more

About the questions to the author
- I looked (not very long though...) for an existing library. ICU deals with localization. The planned usage of this code has nothing to do with localization, also locale ids look a lot like BCP47 language tags but they're not the same thing. So it doesn't look like a good fit to me. It would also be another external dependency.
- nothing changes for a large library. The migration is written so that the number of update queries is the minimum possible (ex. update all records with original language "eng" at once)
- perf measurement: I don't know since there is not much real data yet. As indicated in PR description, more fast paths could be added for the lang-region form or a lang-script form which are probably relatively common.
- Code for IANA registry is mostly ready, though integration with code of this PR is not done yet.

**Comment by CrystalP** [comment]
force-pushed for clang-format

**Comment by CrystalP** [comment]
Rebased on master to merge a couple additional iso 639 unit tests
Commit to remove friendship between CBcp47 and CBcp47Formatter and use public individual subtag accessors for the time being.
The main users of CBcp47 haven't been written yet and the interface they need is not clear yet so may be reversed later in favor of different accessors / private fields.

I believe the PR is ready for review/merge.

**Comment by CrystalP** [comment]
fixed clang-format... & log an error when a tag or description from nfo or scraper is not recognized.

**Comment by CastagnaIT** [comment]
great work i would like test it, i hope in the week

**Comment by CrystalP** [comment]
Jenkins build this please

**Comment by CrystalP** [comment]
Rebased for the DB bump - now 141 since 140 was used by the merged 27566.

File: xbmc/utils/LangCodeExpander.cpp [hunk 1]
```diff
@@ -581,13 +546,70 @@ std::string CLangCodeExpander::ConvertToISO6392T(const std::string& lang)
...
+    return true;
+
+  if (auto tag = CBcp47::ParseTag(code); tag.has_value())
+  {
+    if (tag->IsValid())
+    {
+      tag->Canonicalize();
+
+      if (use == CLangCodeExpander::Bcp47Usage::USAGE_AUDIO)
+        tag->ToAudioLanguageTag();
```
Comment (CastagnaIT) [inline xbmc/utils/LangCodeExpander.cpp]: this kind of distinction about Audio/other types and related LanguageTagSource
used only by a single specific method of CVideoInfoTag for Original lang, and when not internal only,
seem too much specific, IMO should not be done by LangCodeExpander
all this sound as a big workaround

**Comment by CastagnaIT** [comment]
have you removed some changes? i dont see anymore the bcp47 big file of lang codes etc

**Comment by CrystalP** [comment]
> have you removed some changes? i dont see anymore the bcp47 big file of lang codes etc

That's in the other PR 27411 marked WIP. It's slated to go after this one.
The piece that's missing is maybe to load the data async. I don't know how low power devices react.

File: xbmc/utils/LangCodeExpander.cpp (hunk 1 above)
Comment (CrystalP) [reply xbmc/utils/LangCodeExpander.cpp]: I think the interface or functionality of the class CBcp47 won't change much in future PRs but the integration with the rest of the code (including LangCodeExpander) will, I don't know the full picture yet, most of the client code hasn't been spec'ed or written yet.

The existing architecture has LangCodeExpander centralizing all language code conversion functions, and tags are simple text strings . I'm not saying it's a good design but changing that goes beyond this PR. Bcp 47 is one more conversion, so it fits there (for now). Not applying the USAGE_AUDIO filter here would mean another round-trip to parse/filter/format as text.

The purpose of LanguageTagSource is to distinguish text tags that come from the "outside" and require validation from those previously validated and that can be loaded directly, avoiding a parse.
This is likely to change later as I'm expecting that audio/subtitle tracks languages will not be text strings anymore but some language tag object, maybe CBcp47 or maybe something else.
To make things efficient later maybe I'll change the storage in the db to something like a prefix / value, ex.
iso6391: en
bcp47: zh-Hans-HK
user: yue
addon: en-CA
and allow unvalidated strings without prefix to distinguish them, kept to allow a future match with future revision of the  bcp47 registry or new entries in advanced settings or installation of language addon

Comment (CastagnaIT) [reply xbmc/utils/LangCodeExpander.cpp]: about `Bcp47Usage`
im still thinking that is not the best thing todo
i listing my concerns:

1) Ok in the specific of  "script" info its for written text scope, so it would not be suitable for audio, but is it really necessary to implement this filter? Do you know if other players do this? otherwise we can just let it go. because it means that someone has mistakenly or intentionally coded the language with that specific lang code in the audio track for X reason

2) this seem applyed only DB/Python level, if you play a video with e.g. "zh-Hans" this will be shown on GUI as usual, without "script" removal, looks inconsistent

3) if we really need to have an "info removal" feature in the conversion, this should be clearly implemented, e.g. Bcp47Usage --renamed to something like--> Bcp47ConvSetting, where you can specify what keep or remove, or have some sort of selectable info or so...

4) this is applied only to "original" audio, not to all audio track types?

IMO it would probably best to remove `Bcp47Usage` info removal code, so allow each use case to do what they want

Comment (CrystalP) [reply xbmc/utils/LangCodeExpander.cpp]: The audio tag filter is not mandatory and can be postponed. At this point it only fixes tagging errors (the RFC 5646 states clearly that script applies only to written content, see 2.2.3 and 4.1.2).

I was thinking about the player's audio/sub track selection (future PR). Preferred audio/sub can be set to "User interface language" and that could be a problem for users that use a locale with a script part. It would have to be filtered out when used as preferred audio language.

Yes there are missing code paths (your 2. and 4.) because I didn't want to make the PR bigger than it already is.

Other players don't implement any of this from what I've seen so far so comparisons are not valid.

**Comment by CrystalP** [comment]
removed the script subtag filter for audio, kept the code in a commit in case it's needed in some form later on.

File: xbmc/utils/i18n/Bcp47.h [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  // Accessors
+  Bcp47TagType GetType() const { return m_type; }
+  std::string GetLanguage() const { return m_language; }
+  std::vector<std::string> GetExtLangs() const { return m_extLangs; }
+  std::string GetScript() const { return m_script; }
+  std::string GetRegion() const { return m_region; }
+  std::vector<std::string> GetVariants() const { return m_variants; }
+  std::vector<Bcp47Extension> GetExtensions() const { return m_extensions; }
+  std::vector<std::string> GetPrivateUse() const { return m_privateUse; }
+  std::string GetGrandfathered() const { return m_grandfathered; }
```
Comment (CastagnaIT) [inline xbmc/utils/i18n/Bcp47.h]: for all these methods since values are stored in the class
have sense return a const reference to save memory
(with relative code adaptations)

e.g.
`const std::string& GetLanguage() const { return m_language; }`

File: xbmc/utils/LangCodeExpander.cpp (hunk 1 above)
Comment (CastagnaIT) [reply xbmc/utils/LangCodeExpander.cpp]: yeah if can be done separately its better than having an inconsistent partial implementation

**Comment by CastagnaIT** [comment]
the only thing i found are uncomfortable ISO3166_1 names, are really too long,
example:
immagine

and there are even longer names

this will cause considerable confusion in the list, making it difficult to read the tracks
IMO we should consider shortening the names listed (separate PR)

File: xbmc/utils/i18n/Bcp47.h (hunk 2 above)
Comment (CrystalP) [reply xbmc/utils/i18n/Bcp47.h]: I hear you but I'm weary to do this optimization before the rest of the functionality is implemented in other PRs.
Today the class is immutable, but that may change. I'll add a TODO, OK?

edit: it's also likely the accessors will be removed.

**Comment by CrystalP** [comment]
> the only thing i found are uncomfortable ISO3166_1 names, are really too long,

That will be taken care of by the use of the BCP47 registry.

File: xbmc/utils/i18n/Bcp47.h (hunk 2 above)
Comment (CrystalP) [reply xbmc/utils/i18n/Bcp47.h]: though about it again and implemented.

**Comment by CrystalP** [comment]
Had the CBcp47 accessors return const references and adapted a few callers that could take advantage of it.

**Comment by CrystalP** [comment]
no functional changes:
rebased on master for easier build.
Renamed the new TestUtils.h/c.pp files to TestI18nUtils.h/.cpp for clarity in IDE - there already is a TestUtils..h/cpp pair in xbmc/test and it's more logical for them to "own" the name.

**Review by CastagnaIT — APPROVED** [review]
i dont have any other question, from my side its ok,
the only thing that i find hard to understand is the regex parser,
i tried checking it with an AI and it seems to be sufficient

**Comment by CrystalP** [comment]
Yeah the regexp... that was fun,  I was (still am) a regexp newbie, it took some time with an online checker to figure out the basic structure. It covers the full grammar of the RFC except for the grandfathered tags, handled separately.
I added unit tests to prove that it works as expected.
Thanks for the review!

**Comment by CrystalP** [comment]
@claude any final comments?

**Comment by CrystalP** [comment]
@CastagnaIT I made some progress on the bcp47 registry and updated the PR #27411 if you want to give it a try.

It includes additional commits to start taking advantage of the library.
The tag parser (for .nfo, scrapers) accepts all valid language tags (for example the formerly rejected ISO 639-3 yue for Cantonese and numeric countries like 419 of the common es-419)
Formatting the tags as English (for language info labels) uses the registry for all components.

It helps in the case of en-US where US is now mapped to "United States" instead of "United States of America (the)" but is still not ideal in other cases.
For example yue is mapped to "Yue Chinese", although the registry also has the probably preferable "Cantonese", but how to know when to use the first or the nth description... A simplistic heuristic would be to pick the shortest one - happens to work for yue but no idea for the more general case.
The ISO tables may have some preferable English names... I just don't know enough about the subject. Some finetuning may be required later.

**Comment by rockrider69** [comment]
@CrystalP
more of a question here ;) as I am not sure it is fully related to this PR.
I am maintaining an addon called "Language Preference Manager".

From Piers, I see that external subs are now reported in Json call (Player.GetProperties fields) as 2-digits language codes in (ex. en, fr, ...). It was 3-digits up to now.

The internal subtitles streams (mkv, mp4) are still reported as 3-digits codes, unchanged.

Why this difference ? Are the internal streams going to move to 2-digits as well in the future ?

**Comment by CrystalP** [comment]
The implementation is in-between two worlds, I'm not sure I'll be able to finish in v22.
The language of external audio and subs (and inputstream streams) is parsed and returned as BCP47, meaning a 2 letter code when one exists for the language, 3 letter code otherwise. With optional country, script, etc. appended after the language code.

Yes the plan is to move everything to BCP47. With a BCP47-aware comparison function, to match user languages with the stream languages.
ffmpeg is a roadblock though as it cannot read the BCP47 tag from mkv or mp4 for example and retrieves only the backwards compatible iso659.

**Comment by rockrider69** [comment]
@CrystalP 
thx a lot for the answer. Looks like I need to make a plan for progressive BCP47 support as well ;)
